Dev/tlvf fix dynamic list

tomereli opened the pull request:
Multiple bug fixes for dynamic lists in tlvf

- Fix potential null pointer dereference when calling class_swap()
when the pointer is uninitialized.
- fix int length lists parse bugs
- add length function for unknown length lists

Commit:
tlvf: fix potential null pointer dereference

Fix potential null pointer dereference when calling class_swap()
when the pointer is uninitialized.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: fix int length lists parse bugs

Integer length lists are fixed size lists, for example:

mac:
  _type: uint8_t
  _length: [6]

In both cases, on RX we do parsing, so the values of the length and the
actual list is already set, so:

- Fix bug when int length list length variable value is incremented on
parse
- Fix bug when int length list values are initialized on parse

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: add length function for unknown length lists

- Add length function for unknown length lists
- Add unknown length list to the inner class in tlvVarList.yaml
- and a test in tlvf test to verfiy length function works as expected

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -1007,10 +1007,6 @@ def addClassVarLenMethods(self, obj_meta, param_type, param_name, param_meta, pa
             lines_cpp.append( "%sreturn false;" % self.getIndentation(2))
             lines_cpp.append( "%s}" % self.getIndentation(1) )
 
-            lines_cpp.append( "%sif (ptr->getStartBuffPtr() != (uint8_t *)m_%s) {" % (self.getIndentation(1), param_name) )
```
Why did you remove it? Moran's change fixes the bug in this code

tomereli replied:
I removed it since it is not correct, parsing fails for list of classes of multiple entries. Once Moran's change is merged, I'll rebase and we'll get it back.

tomereli replied:
Brought it back

Commit:
tlvf: fix unknown length lists parsing

When parsing an unknown length list, there is a sanity test that verfies
the add list member start pointer matches the expected.
However, this is done only for created TLVs and not parsed ones, so
the test fails on parse.
Fix it by updating tlvf.py to always update the src pointer to the last
added list member.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on framework/tlvf/tlvf.py:
```diff
@@ -851,10 +851,11 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
-                                 "%sm_%s[i] = %s;"% (self.getIndentation(1), param_name, str(param_val) if param_val != None else str(param_val_const)),
-                                 "}",
-                                 ""]
+                    lines_cpp += ["if (!m_parse__) {",
```
It would be clearer both for the generated code and the code generator itself if the parse condition was there only once. So for tlvf, that means to swap the conditions `if (param_val_const or param_val)` and `if is_int_len or is_const_len` and then to merge the two int/const len conditions that follow each other, and then to move the generated `if (!m_parse__)` to the beginning of that condition. Am I making sense?

(nitpick).

arnout commented on framework/tlvf/yaml/tlvf/test/tlvVarList.yaml:
```diff
@@ -38,3 +38,7 @@ cInner:
     _type: uint8_t
     _length: [ list_length ]
   var1: uint32_t
+  unknown_length_list:
```
Could you give a different name than the unkown_length_list that already exists in the outer class? Otherwise the code gets a little bit confusing.

The same problem already exists for var1, with things like `var1()->var1()` in the code...

Commit:
tlvf: add test for int length lists

Add a test for int length lists using tlvMacAddress.yaml.

arnout commented on framework/tlvf/yaml/tlvf/test/tlvVarList.yaml:
```diff
@@ -38,3 +38,7 @@ cInner:
   var1: uint32_t
+  unknown_length_list:
+    _type: uint8_t
+    _length: []
```
This cannot possibly be parsed. An unknown length list can only exist as the last field in a TLV, because it uses the TLV length to determine the length of the list. Here, cInner is used in e.g. the outer class's var1, so there is no way where the boundary lies between var1's unknown_length_list and the outer unkown_length_list. (Note the confusion with the names :-).

arnout reviewed (changes requested):
Commit 1-3 are good to go, except that the wrapping of commit 3's message is not correct.

I would like to see a test for the issues fixed in commit 1-3. Preferably as a new commit that comes before the rest, so we see the test break in the current situation. For the first one, we can just reuse the existing tlvMacAddress.

Commit 4 needs to be fixed.

tomereli commented:
> Commit 1-3 are good to go, except that the wrapping of commit 3's message is not correct.
> 
> I would like to see a test for the issues fixed in commit 1-3. Preferably as a new commit that comes before the rest, so we see the test break in the current situation. For the first one, we can just reuse the existing tlvMacAddress.
> 
> Commit 4 needs to be fixed.

Commit 1 - no test I can think of since it doesn't really happen, it is just safer to check the pointer we are actually using.
Commit 2 - added a test
Commit 3 - Just adding a length getter function, no test can be added.

tomereli commented on tools/prplMesh.code-workspace:
```diff
@@ -77,7 +77,10 @@
 			"atomic": "cpp",
 			"codecvt": "cpp",
-			"memory": "cpp"
+			"memory": "cpp",
```
Not related to the PR - Please ignore

tomereli commented:
> Commit 1-3 are good to go, except that the wrapping of commit 3's message is not correct.
> 
> I would like to see a test for the issues fixed in commit 1-3. Preferably as a new commit that comes before the rest, so we see the test break in the current situation. For the first one, we can just reuse the existing tlvMacAddress.
> 
> Commit 4 needs to be fixed.

I opened a new PR  #144 for commits 1-3 - please review, then i'll rebase and fix this one.

tomereli commented:
Fixed in #148
